### PR TITLE
Feat: Redis Database Driver - updated

### LINKED ISSUE #29: Storage - Redis driver

### PR DESCRIPTION
**Pull request by Man-RP**
Hey,
I've added the option to use a Redis DB server.
Its base on #163 and closes #29 .

### CODE REVIEW

File: .env.development [hunk 1]
```diff
@@ -4,17 +4,23 @@ BASE_REDIRECT_URL=http://localhost:3000/u/
 LOG_LEVEL=debug
 npm_package_type=module
 
-########## Auth ##########
+# ######### Auth ##########
```
Comment (SnirShechter) [inline .env.development:7]:
```suggestion
########## Auth ##########
```

File: .env.development [hunk 2]
```diff
@@ -4,17 +4,23 @@ BASE_REDIRECT_URL=http://localhost:3000/u/
...
 AUTH_DRIVER=BearerToken
 AUTH_BEARER_TOKEN=123
 
-########## Storage ##########
-## InMemory/Postgres/Sqlite
-STORAGE_DRIVER=Sqlite
+# ######### Storage ##########
+# # InMemory/Postgres/Sqlite/Redis
+STORAGE_DRIVER=Redis
```
Comment (SnirShechter) [inline .env.development:13]:
```suggestion
########## Storage ##########
## InMemory/Postgres/Sqlite/Redis
STORAGE_DRIVER=Sqlite
```

File: .env.development [hunk 3]
```diff
@@ -4,17 +4,23 @@ BASE_REDIRECT_URL=http://localhost:3000/u/
...
 
-#### Postgres Config
-## To run a local database suitable for this configuration, see the contribution guide
+# ### Postgres Config
+# # To run a local database suitable for this configuration, see the contribution guide
```
Comment (SnirShechter) [inline .env.development:16]:
```suggestion
#### Postgres Config
## To run a local database suitable for this configuration, see the contribution guide
```

File: .env.development [hunk 4]
```diff
@@ -4,17 +4,23 @@ BASE_REDIRECT_URL=http://localhost:3000/u/
...
 # POSTGRES_STORAGE_HOST=localhost
 # POSTGRES_STORAGE_USER=postgres
 # POSTGRES_STORAGE_PASSWORD=postgres
 # POSTGRES_STORAGE_DATABASE=postgres
+
+# ## Redis Config
+ REDIS_STORAGE_HOST = localhost
+ REDIS_STORAGE_PORT = 6379
+ REDIS_STORAGE_USERNAME = redis
+ REDIS_STORAGE_PASSWORD = redis
```
Comment (SnirShechter) [inline .env.development:26]:
```suggestion
### Redis Config
# REDIS_STORAGE_HOST = localhost
# REDIS_STORAGE_PORT = 6379
# REDIS_STORAGE_USERNAME = redis
# REDIS_STORAGE_PASSWORD = redis

```

File: src/config/normalize.ts [hunk 5]
```diff
@@ -66,6 +66,15 @@ function getStorageDriverConfig(
 			driverConfig: storage.sqliteDriverConfig,
 			...extras,
 		}
+	} else if (storage.driverName === Redis) {
+		return {
+			driverName: Redis,
+			driverConfig: {
+				...storage.redisDriverConfig,
+				port: Number(storage.redisDriverConfig.port),
```
Comment (SnirShechter) [inline src/config/normalize.ts:74]:
```suggestion
				port: +storage.redisDriverConfig.port,
```

File: src/services/storage/drivers/redis/index.ts [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+	}
+
+	async shutdown(): Promise<void> {
+		await this._client.disconnect()
+	}
+
+	initialize(): Promise<void> {
+		//connection with the db checked in storage/index runWithRetries
+		return new Promise<void>((resolve) => resolve())
+	}
```
Comment (SnirShechter) [inline src/services/storage/drivers/redis/index.ts:27]:
```suggestion
	async initialize(): Promise<void> {
		// Connection with the DBis already checked in storage/index using runWithRetries
		return
	}
```

File: src/services/storage/drivers/redis/index.ts [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+
+	url = new (class RedisUrlStorage {
+		constructor(public storage: RedisStorage) {}
+
+		tempPropertyMap = new Map()
```
Comment (SnirShechter) [inline src/services/storage/drivers/redis/index.ts:32]: Why do we need the tempPropertyMap?

File: src/services/storage/drivers/redis/index.ts [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+					createdAt: this.tempPropertyMap.get('createdAt'),
+					updatedAt: this.tempPropertyMap.get('updatedAt'),
+				}
+		}
+
+		fillTempPropertyMap(key: string, value: string): void {
+			this.tempPropertyMap.set(key, value)
+		}
+
+		processUrlFromStorage(rawUrlData: string): void {
```
Comment (SnirShechter) [inline src/services/storage/drivers/redis/index.ts:132]: Why do we need this function?

File: src/services/storage/drivers/redis/index.ts [hunk 9]
```diff
@@ -0,0 +1,138 @@
...
+			else
+				return {
+					id: this.tempPropertyMap.get('id'),
+					url: this.tempPropertyMap.get('url'),
+					createdAt: this.tempPropertyMap.get('createdAt'),
+					updatedAt: this.tempPropertyMap.get('updatedAt'),
+				}
+		}
+
+		fillTempPropertyMap(key: string, value: string): void {
```
Comment (SnirShechter) [inline src/services/storage/drivers/redis/index.ts:128]: Why do we need this function?

File: src/services/storage/drivers/redis/index.ts [hunk 10]
```diff
@@ -0,0 +1,138 @@
...
+			)
+			return storedUrlWithInfo
+		}
+
+		async fetchUrlInfoFromDB(id: string, withInfo: boolean): Promise<UrlWithInformation | StoredUrl> {
+			this.tempPropertyMap.clear()
+			const rawUrlData = (await this.storage.client.get(id)) as string
+			if (!rawUrlData) throw new NotFoundError()
+			this.processUrlFromStorage(rawUrlData)
+			if (withInfo)
```
Comment (SnirShechter) [inline src/services/storage/drivers/redis/index.ts:108]: Multiline if/else should be written with curly braces

File: src/services/storage/drivers/redis/index.ts [hunk 11]
```diff
@@ -0,0 +1,138 @@
...
+			}
+			await this.storage.client.setex(
+				storedUrlWithInfo.id,
+				config.url.lifetimeMs / 1000, //TODO: set ttl according to _config flag
+				JSON.stringify(storedUrlWithInfo),
+			)
+			return storedUrlWithInfo
+		}
+
+		async fetchUrlInfoFromDB(id: string, withInfo: boolean): Promise<UrlWithInformation | StoredUrl> {
```
Comment (SnirShechter) [inline src/services/storage/drivers/redis/index.ts:103]: This function is redundant (see the `get` function)

File: src/services/storage/drivers/redis/index.ts [hunk 12]
```diff
@@ -0,0 +1,138 @@
...
+	}
+
+	url = new (class RedisUrlStorage {
+		constructor(public storage: RedisStorage) {}
+
+		tempPropertyMap = new Map()
+
+		async get(id: string, options: { withInfo: boolean }): Promise<StoredUrl | UrlWithInformation> {
+			return await this.fetchUrlInfoFromDB(id, options.withInfo)
+		}
```
Comment (SnirShechter) [inline src/services/storage/drivers/redis/index.ts:36]:
```suggestion
		async get(id: string, options: { withInfo: boolean }): Promise<StoredUrl | UrlWithInformation> {
			const rawUrlData = (await this.storage.client.get(id)) as string
			if (!rawUrlData) throw new NotFoundError()
			
			const urlData = JSON.parse(rawUrlData) as UrlWithInformation

			if (withInfo) return urlData
			else {
				return {
					id: urlData.id,
					url: urlData.url,
					createdAt: urlData.createdAt,
					updatedAt: urlData.updatedAt,
				}
			}
		}
```

File: src/services/storage/drivers/redis/index.ts [hunk 13]
```diff
@@ -0,0 +1,138 @@
...
+			// Not in use since Redis has build-in mechanism to delete keys after period of time
+			return -1
+		}
+
+		async edit(id: string, url: string): Promise<StoredUrl> {
+			const dbUrl: StoredUrl = await this.fetchUrlInfoFromDB(id, false)
+			dbUrl.updatedAt = new Date().toISOString()
+			dbUrl.url = url
+			const config = getConfig()
+			//TODO: set ttl according to _config flag
```
Comment (SnirShechter) [inline src/services/storage/drivers/redis/index.ts:52]: Go over and fix/remove all TODOs

File: src/services/storage/drivers/redis/index.ts [hunk 14]
```diff
@@ -0,0 +1,138 @@
+import IORedis from 'ioredis'
+import { StorageDriver } from '../../types'
+import { RedisStorageConfig } from '../../types/config'
+import { StoredUrl, UrlRequestData, UrlWithInformation } from '../../types/url'
+import cryptoRandomString from 'crypto-random-string'
+import { getConfig } from '../../../../config'
```
Comment (SnirShechter) [inline src/services/storage/drivers/redis/index.ts:6]: You shouldn't import getConfig. Every configuration you need should already be in RedisStorageConfig

File: src/services/storage/drivers/redis/index.ts [hunk 15]
```diff
@@ -0,0 +1,138 @@
...
+			urlFromDb.infoVisitCount = ++urlFromDb.infoVisitCount
+
+			const ttl = await this.storage.client.ttl(id)
+			await this.storage.client.setex(id, ttl, JSON.stringify(urlFromDb))
+		}
+
+		async incVisitCount(id: string): Promise<void> {
+			if (!id) throw new OperationFailed('Must provide an id')
+
+			const urlFromDb: UrlWithInformation = <UrlWithInformation>await this.fetchUrlInfoFromDB(id, true)
```
Comment (SnirShechter) [inline src/services/storage/drivers/redis/index.ts:72]:
```suggestion
			const urlFromDb: UrlWithInformation = <UrlWithInformation>await this.get(id, true)
```

File: src/services/storage/drivers/redis/index.ts [hunk 16]
```diff
@@ -0,0 +1,138 @@
...
+			const config = getConfig()
+			//TODO: set ttl according to _config flag
+			await this.storage.client.setex(dbUrl.id, config.url.lifetimeMs / 1000, JSON.stringify(dbUrl))
+			return dbUrl
+		}
+
+		async incInfoCount(id: string): Promise<void> {
+			if (!id) throw new OperationFailed('Must provide an id')
+
+			const urlFromDb: UrlWithInformation = <UrlWithInformation>await this.fetchUrlInfoFromDB(id, true)
```
Comment (SnirShechter) [inline src/services/storage/drivers/redis/index.ts:60]:
```suggestion
			const urlFromDb: UrlWithInformation = <UrlWithInformation>await this.get(id, true)
```

File: src/services/storage/drivers/redis/index.ts [hunk 17]
```diff
@@ -0,0 +1,138 @@
...
+			return this.storage.client.del(id)
+		}
+
+		async deleteOverdue(timespanMs: number): Promise<number> {
+			// Not in use since Redis has build-in mechanism to delete keys after period of time
+			return -1
+		}
+
+		async edit(id: string, url: string): Promise<StoredUrl> {
+			const dbUrl: StoredUrl = await this.fetchUrlInfoFromDB(id, false)
```
Comment (SnirShechter) [inline src/services/storage/drivers/redis/index.ts:48]:
```suggestion
			const dbUrl: StoredUrl = await this.get(id, false)
```

**Review by SnirShechter — CHANGES_REQUESTED** [review]

File: src/config/normalize.ts (hunk 5 above)
Comment (Man-RP) [reply src/config/normalize.ts:74]: I had to add casting.
Without it i'm getting a typescript error.

File: src/services/storage/drivers/redis/index.ts [hunk 18]
```diff
@@ -0,0 +1,109 @@
...
+	get client(): IORedis.Redis {
+		return this._client
+	}
+
+	async shutdown(): Promise<void> {
+		await this._client.disconnect()
+	}
+
+	async initialize(): Promise<void> {
+		// Connection with the DB is already checked in storage/index using runWithRetries
```
Comment (SnirShechter) [inline src/services/storage/drivers/redis/index.ts:24]: That was actually not true

**Review by SnirShechter — APPROVED** [review]

**Pull request merged**